### ISSUE TITLE
API Proposal: ServiceBase.IsRunningInWindowsService

### ISSUE THREAD

**Issue by davidfowl** [dotnet/corefx#36842]
We recently added an API to detect if a .NET process was being lauched as a windows service (https://github.com/aspnet/Extensions/pull/1369). I believe this belongs in SeviceBase as it would be useful to have this feature outside of our hosting wrapper.

```C#
public class ServiceBase
{
    public static bool IsRunningInWindowsService { get; }
}
```

I'm open to other names.
- IsRunningUnderWindowsService

cc @danmosemsft

**Comment by anurse** [comment]
Example implementation: https://github.com/aspnet/Extensions/pull/1369/files

**Comment by danmosemsft** [comment]
@anipik owns this area. Why is the API useful?

**Comment by davidfowl** [comment]
We use it in our hosting layer to make the windows service specific logic noop. The benefit being that we can light up specific behavior when running in that context (like hooking into the lifetime events). This means my application runs in both modes (service or self hosted) seamlessly.

We’ve also had a customer ask us for this logic because they wanted a change to write some service specific logic in their startup.

**Comment by danmosemsft** [comment]
@anurse is that method documented, eg there is some reason to believe the service host process name won't change in future Windows versions?

I don't feel strongly but Im inclined to think we would want more evidence of need. If it was wrapping something like `::IsAService()` it would be a lower bar in my mind.

**Comment by davidfowl** [comment]
I just described how we use it. What evidence are you looking for? Are you not convinced that there are scenarios where knowing that you are running in a windows service is useful?

If windows changes it in the future we’ll update the code. If you know of a better way to detect that scenario there’ll be no pushback

**Comment by Wraith2** [comment]
I believe this would be useful. In the past I've hooked up a log reader if running outside a service so I can direct the output to a user interface.

**Comment by danmosemsft** [comment]
if this was on ServiceBase the app would require a reference to the compat pack which it might not otherwise need. On the flip side, it is very Windows specific...

**Comment by davidfowl** [comment]
@danmosemsft do yo have a proposal? Where would it live?

**Comment by anurse** [comment]
There is [Environment.UserInteractive](https://docs.microsoft.com/en-us/dotnet/api/system.environment.userinteractive?redirectedfrom=MSDN&view=netframework-4.7.2#System_Environment_UserInteractive) which checks if the application is running in an "Interactive Session".

From the docs: 

> The UserInteractive property reports false for a Windows process or a service like IIS that runs without a user interface. If this property is false, do not display modal dialogs or message boxes because there is no graphical user interface for the user to interact with.

In my initial cursory testing, a console app reports `Environment.UserInteractive` as **`true`**.

We might need further testing here, but it *may* serve our purpose. It's not guaranteed to indicate that you're in a Windows Service though since there may be other non-Windows Service non-Interactive scenarios.

**Comment by Tratcher** [comment]
That would also include scheduled tasks.

**Comment by anurse** [comment]
That's what I was looking for :). I figured there might be a counter-example here.

**Comment by davidfowl** [comment]
Doesn't that also report true when running in IIS?

**Comment by danmosemsft** [comment]
As an aside, I had a quick look at the service related Win32 API and I didn't see an alternative way to check the current process is a service. The API's all deal with handles or service names (from which you get the handle). I don't see a way eg to enumerate services from a process handle.

**Comment by MarcoRossignoli** [comment]
Name+path(`%SystemRoot%\System32\services.exe`)+`Environment.UserInteractive` ?

**Comment by MarcoRossignoli** [comment]
Another idea could be check associated windows station name, from `Windows Internals 6`:
```
Unless other directed, the Windows subsystem associates services running in the local system account with a nonvisible windows station named Service-0x0-3e7$ that all noninteractive services share.
...
However only processes owned by the system and Windows services run in session 0; all other logon sessions, including those of console users, run in different session. Any windows displayed by process in session 0 is therefore not visible to the user.
```
I did some check and it works also under IIS

![image](https://user-images.githubusercontent.com/7894084/56240922-e54d8580-6094-11e9-949f-06fb7cd6ccfc.png)

The only issue is in case of `SERVICE_INTERACTIVE_PROCESS` flag for LocalSystem account...process will be associated with WinSta0, this is however not recommended for security reason and supported for backward compatibility.

**Comment by danmosemsft** [comment]
@Anipik @maryamariyan please limit 3.0 milestone to issues that must be fixed to ship 3.0.

**Comment by terrajobst** [comment]
Conclusion

* The location makes sense because if the API returns true, consumers are likely ending up using APIs on `ServiceBase`
* We should consider shortening the name to `IsRunningAsService` as Windows is implied (the entire type is Windows-specific)
* We can still expose a platform-agnostic API (that detects, say, Windows services and Linux daemons) but that's out of scope for this scenario

**Comment by george-chakhidze** [comment]
`Environment.UserInteractive` always returns `true` in .NET Core. Maybe, while at it, you could properly implement this API as well?

**Comment by danmosemsft** [comment]
@george-chakhidze yep, this was done for 5.0. And as was pointed out above, it is not the same as this proposed API, although for some use cases it may be enough.
